ts/client/database/databaseEngine.ts

```diff
@@ -1,12 +1,107 @@
-import { Hashable } from "../../interfaces";
-import { Leaf, LeafFactoryFunc } from "../../tree/leaves/Leaf";
-import { MemoryEngine } from "../storageEngine/memoryEngine";
+import { ZERO_BYTES32 } from "../../constants";
+import { TreeAtLevelIsFull } from "../../exceptions";
+import { Hashable, Vacant } from "../../interfaces";
+import { Hasher } from "../../tree";
+import { DBTree } from "../../tree/dbTree";
+import { LeafFactoryFunc } from "../../tree/leaves/Leaf";
+import { StorageEngine, WithWitness } from "../storageEngine/interfaces";
 
-export class DatabaseEngine<Item extends Hashable> extends MemoryEngine<Item> {
-    private readonly factory: LeafFactoryFunc<Leaf<Item>>;
+export class DatabaseEngine<Item extends Hashable>
+    implements StorageEngine<Item> {
+    private tree: DBTree;
+    public readonly leafFactory: LeafFactoryFunc<Item>;
 
-    constructor(depth: number, factory: LeafFactoryFunc<Leaf<Item>>) {
-        super(depth);
-        this.factory = factory;
+    constructor(depth: number, factory: LeafFactoryFunc<Item>) {
+        this.tree = DBTree.new(
+            depth,
+            factory.name,
+            Hasher.new("bytes", ZERO_BYTES32)
+        );
+        this.leafFactory = factory;
+    }
+
+    private checkSize(itemID: number) {
+        if (itemID >= this.tree.setSize)
+            throw new Error(
+                `Want itemID ${itemID} but the tree has only ${this.tree.setSize} leaves`
+            );
+    }
+
+    public get root() {
+        return this.tree.root;
+    }
+
+    public async get(itemID: number): Promise<Item> {
+        this.checkSize(itemID);
+        try {
+            return (await this.leafFactory.fromDB(itemID)).item;
+        } catch (error) {
+            throw new Error(`Item not exists  itemID: ${itemID}`);
+        }
```

Comment:
Is this `try/catch` needed, or will the underlying leveldb `get` in `leafFactory.fromDB` throw? If `get` can return an undefined value, this would be a better not found approach:
```suggestion
        const leaf  = await this.leafFactory.fromDB(itemID);
        if (!leaf) {
            throw new Error(`Item not exists  itemID: ${itemID}`);
        }
        return leaf.item;
```

Reply:
yeah, get throws an error if the key is not present in the db. 
`NotFoundError: Key not found in database`